File: src/main/java/com/paragon/client/systems/module/impl/combat/Surround.java
package com.paragon.client.systems.module.impl.combat;

import com.paragon.Paragon;
import com.paragon.api.util.player.InventoryUtil;
import com.paragon.api.util.player.RotationUtil;
import com.paragon.api.util.render.ColourUtil;
import com.paragon.api.util.render.RenderUtil;
import com.paragon.api.util.world.BlockUtil;
import com.paragon.client.managers.rotation.Rotate;
import com.paragon.api.module.Module;
import com.paragon.api.module.Category;
import com.paragon.api.setting.Setting;
import net.minecraft.entity.Entity;
import net.minecraft.init.Blocks;
import net.minecraft.network.play.client.CPacketEntityAction;
import net.minecraft.network.play.client.CPacketPlayer;
import net.minecraft.util.EnumFacing;
import net.minecraft.util.EnumHand;
import net.minecraft.util.math.*;
import net.minecraft.util.text.TextFormatting;

import java.awt.*;
import java.util.HashMap;
import java.util.Map;

/**
 * @author Wolfsurge
 * @since 08/05/2022
 */
public class Surround extends Module {

    public static Surround INSTANCE;

    // General settings
    public static Setting<Disable> disable = new Setting<>("Disable", Disable.NEVER)
            .setDescription("When to automatically disable the module");

    public static Setting<Double> blocksPerTick = new Setting<>("BlocksPerTick", 4D, 1D, 10D, 1D)
            .setDescription("The maximum amount of blocks to be placed per tick");

    public static Setting<Center> center = new Setting<>("Center", Center.MOTION)
            .setDescription("Center the player on the block when enabled");

    public static Setting<Air> air = new Setting<>("Air", Air.SUPPORT)
            .setDescription("Place blocks beneath where we are placing");

    // Rotate settings
    public static Setting<Rotate> rotate = new Setting<>("Rotate", Rotate.LEGIT)
            .setDescription("How to rotate the player");

    public static Setting<Boolean> rotateBack = new Setting<>("RotateBack", true)
            .setDescription("Rotate the player back to their original rotation")
            .setParentSetting(rotate);

    // Render
    public static Setting<Boolean> render = new Setting<>("Render", true)
            .setDescription("Render a highlight on the positions we need to place blocks at");

    public static Setting<Color> renderColour = new Setting<>("Colour", new Color(185, 17, 255, 130))
            .setDescription("The colour of the highlight")
            .setParentSetting(render);

    // Map of blocks to render
    private Map<BlockPos, EnumFacing> renderBlocks = new HashMap<>();

    public Surround() {
        super("Surround", Category.COMBAT, "Places obsidian around you to protect you from crystals");

        INSTANCE = this;
    }

    @Override
    public void onEnable() {
        if (nullCheck()) {
            return;
        }

        // No obsidian to place
        if (InventoryUtil.getHotbarBlockSlot(Blocks.OBSIDIAN) == -1) {
            return;
        }

        // Center player
        switch (center.getValue()) {
            case MOTION:
                // Set player's motion to walk to the center of the block
                mc.player.motionX = ((MathHelper.floor(mc.player.posX) + 0.5D) - mc.player.posX) / 2;
                mc.player.motionZ = ((MathHelper.floor(mc.player.posZ) + 0.5D) - mc.player.posZ) / 2;
                break;

            case SNAP:
                // Send movement packet
                mc.player.connection.sendPacket(new CPacketPlayer.Position(MathHelper.floor(mc.player.posX) + 0.5, mc.player.posY, MathHelper.floor(mc.player.posZ) + 0.5, mc.player.onGround));

                // Set position client-side
                mc.player.setPosition(MathHelper.floor(mc.player.posX) + 0.5, mc.player.posY, MathHelper.floor(mc.player.posZ) + 0.5);
                break;
        }

    }

    @Override
    public void onTick() {
        if (nullCheck()) {
            return;
        }

        // Check we are in air
        if (disable.getValue().equals(Disable.AIR) && !mc.player.onGround) {
            toggle();
            return;
        }

        // We don't have obsidian in our hotbar
        if (InventoryUtil.getHotbarBlockSlot(Blocks.OBSIDIAN) == -1) {
            Paragon.INSTANCE.getCommandManager().sendClientMessage(TextFormatting.RED + "No obsidian available, Surround disabled!", false);
            toggle();
            return;
        }

        // Blocks we need to place
        Map<BlockPos, EnumFacing> blocks = new HashMap<>();

        // The player's position
        BlockPos playerPos = new BlockPos(MathHelper.floor(mc.player.posX), MathHelper.floor(mc.player.posY), MathHelper.floor(mc.player.posZ));

        for (int x = -1; x <= 1; x++) {
            // Our X position
            if (x == 0) {
                continue;
            }

            // Get position
            BlockPos pos = playerPos.add(x, 0, 0);

            // Check if we need to support the block
            if (canPlaceBlock(pos.down()) && air.getValue().equals(Air.SUPPORT)) {
                // Get offset
                for (EnumFacing facing : EnumFacing.values()) {
                    // Make sure we can place on offset
                    if (BlockUtil.getBlockAtPos(pos.down().offset(facing)) != Blocks.AIR) {
                        // Add block
                        blocks.put(pos.down(), facing);
                        break;
                    }
                }
            }

            // We can place the block or we are supporting it
            if (canPlaceBlock(pos) || air.getValue().equals(Air.SUPPORT) && canPlaceBlock(pos.down())) {
                // Get offset
                for (EnumFacing facing : EnumFacing.values()) {
                    // Make sure we can place on offset
                    if (BlockUtil.getBlockAtPos(pos.offset(facing)) != Blocks.AIR) {
                        // Add block
                        blocks.put(pos, facing);
                        break;
                    }
                }
            }

            for (int z = -1; z <= 1; z++) {
                // Our Z position
                if (z == 0) {
                    continue;
                }

                // Get position
                pos = playerPos.add(0, 0, z);

                // Check if we need to support the block
                if (canPlaceBlock(pos.down()) && air.getValue().equals(Air.SUPPORT)) {
                    // Get offset
                    for (EnumFacing facing : EnumFacing.values()) {
                        // Make sure we can place on offset
                        if (BlockUtil.getBlockAtPos(pos.down().offset(facing)) != Blocks.AIR) {
                            // Add block
                            blocks.put(pos.down(), facing);
                            break;
                        }
                    }
                }

                // We can place the block or we are supporting it
                if (canPlaceBlock(pos) || air.getValue().equals(Air.SUPPORT) && canPlaceBlock(pos.down())) {
                    // Get offset
                    for (EnumFacing facing : EnumFacing.values()) {
                        // Make sure we can place on offset
                        if (BlockUtil.getBlockAtPos(pos.offset(facing)) != Blocks.AIR) {
                            // Add block
                            blocks.put(pos, facing);
                            break;
                        }
                    }
                }
            }
        }

        // Set render blocks
        this.renderBlocks = blocks;

        // Get our original rotation
        Vec2f originalRotation = new Vec2f(mc.player.rotationYaw, mc.player.rotationPitch);

        // Place blocks until either:
        // A. We run out of blocks
        // B. We have placed the maximum amount of blocks for this tick
        for (double i = 0; i < blocksPerTick.getValue() && i < blocks.size(); i += 1) {
            // Get position
            BlockPos pos = blocks.keySet().toArray(new BlockPos[0])[(int) i];

            // Get facing
            EnumFacing facing = blocks.get(pos);

            // Get rotation
            Vec2f rotation = RotationUtil.getRotationToBlockPos(pos, 0.5);

            // Rotate to position
            if (rotate.getValue().equals(Rotate.LEGIT)) {
                mc.player.rotationYaw = rotation.x;
                mc.player.rotationPitch = rotation.y;
                mc.player.rotationYawHead = rotation.x;
            } else if (rotate.getValue().equals(Rotate.PACKET)) {
                mc.player.connection.sendPacket(new CPacketPlayer.Rotation(rotation.x, rotation.y, mc.player.onGround));
            }

            // Get current item
            int slot = mc.player.inventory.currentItem;

            // Slot to switch to
            int obsidianSlot = InventoryUtil.getHotbarBlockSlot(Blocks.OBSIDIAN);

            if (obsidianSlot != -1) {
                // Switch
                mc.player.inventory.currentItem = obsidianSlot;

                // Make the server think we are crouching, so we can place on interactable blocks (e.g. chests, furnaces, etc.)
                mc.player.connection.sendPacket(new CPacketEntityAction(mc.player, CPacketEntityAction.Action.START_SNEAKING));

                // Place block
                mc.playerController.processRightClickBlock(mc.player, mc.world, pos.offset(facing), facing.getOpposite(), new Vec3d(pos), EnumHand.MAIN_HAND);

                // Swing hand
                mc.player.swingArm(EnumHand.MAIN_HAND);

                // Stop crouching
                mc.player.connection.sendPacket(new CPacketEntityAction(mc.player, CPacketEntityAction.Action.STOP_SNEAKING));
            }

            // Reset slot to our original slot
            mc.player.inventory.currentItem = slot;
        }

        // Rotate back
        if (rotateBack.getValue()) {
            if (rotate.getValue().equals(Rotate.LEGIT)) {
                mc.player.rotationYaw = originalRotation.x;
                mc.player.rotationPitch = originalRotation.y;
                mc.player.rotationYawHead = originalRotation.x;
            } else if (rotate.getValue().equals(Rotate.PACKET)) {
                mc.player.connection.sendPacket(new CPacketPlayer.Rotation(originalRotation.x, originalRotation.y, mc.player.onGround));
            }
        }

        // Toggle if we have placed all the blocks
        if (this.renderBlocks.isEmpty() && disable.getValue().equals(Disable.FINISHED)) {
            toggle();
        }
    }

    @Override
    public void onRender3D() {
        // We want to render
        if (render.getValue()) {
            renderBlocks.forEach((pos, facing) -> {
                // Render highlight
                RenderUtil.drawFilledBox(BlockUtil.getBlockBox(pos), renderColour.getValue());
                RenderUtil.drawBoundingBox(BlockUtil.getBlockBox(pos), 1, ColourUtil.integrateAlpha(renderColour.getValue(), 255));
            });
        }
    }

    public boolean canPlaceBlock(BlockPos pos) {
        // Iterate through entities in the block above
        for (Entity entity : mc.world.getEntitiesWithinAABB(Entity.class, new AxisAlignedBB(pos))) {
            // If the entity is dead, continue
            if (entity.isDead) {
                continue;
            }

            return false;
        }

        // Block is air
        if (BlockUtil.getBlockAtPos(pos) != Blocks.AIR) {
            return false;
        }

        // There is a block we can place on
        return BlockUtil.getBlockAtPos(pos.down()) != Blocks.AIR ||
                BlockUtil.getBlockAtPos(pos.up()) != Blocks.AIR ||
                BlockUtil.getBlockAtPos(pos.north()) != Blocks.AIR ||
                BlockUtil.getBlockAtPos(pos.south()) != Blocks.AIR ||
                BlockUtil.getBlockAtPos(pos.east()) != Blocks.AIR ||
                BlockUtil.getBlockAtPos(pos.west()) != Blocks.AIR;
    }

    public enum Disable {
        /**
         * Disable once all blocks have been placed
         */
        FINISHED,

        /**
         * Never disable
         */
        NEVER,

        /**
         * Disable when the player is in air
         */
        AIR
    }

    public enum Center {
        /**
         * Move the player to the center
         */
        MOTION,

        /**
         * Snap the player to the center
         */
        SNAP,

        /**
         * Don't center the player
         */
        OFF
    }

    public enum Air {
        /**
         * Place blocks below air blocks, so we can place on top of them
         */
        SUPPORT,

        /**
         * Do not place on air blocks
         */
        IGNORE
    }

}

File: src/main/kotlin/com/paragon/client/systems/module/impl/combat/Surround.kt
package com.paragon.client.systems.module.impl.combat

import com.paragon.Paragon
import com.paragon.api.event.network.PacketEvent
import com.paragon.api.module.Category
import com.paragon.api.module.Module
import com.paragon.api.setting.Setting
import com.paragon.api.util.anyNull
import com.paragon.api.util.player.InventoryUtil
import com.paragon.api.util.player.PlacementUtil
import com.paragon.api.util.player.RotationUtil
import com.paragon.api.util.render.ColourUtil.integrateAlpha
import com.paragon.api.util.render.builder.RenderBuilder
import com.paragon.api.util.world.BlockUtil
import com.paragon.api.util.world.BlockUtil.getBlockAtPos
import com.paragon.bus.listener.Listener
import com.paragon.client.managers.notifications.Notification
import com.paragon.client.managers.notifications.NotificationType
import com.paragon.client.managers.rotation.Rotate
import com.paragon.client.managers.rotation.Rotation
import com.paragon.client.managers.rotation.RotationPriority
import net.minecraft.init.Blocks
import net.minecraft.network.play.client.CPacketEntityAction
import net.minecraft.network.play.client.CPacketPlayer
import net.minecraft.network.play.server.SPacketBlockChange
import net.minecraft.util.EnumFacing
import net.minecraft.util.EnumHand
import net.minecraft.util.math.BlockPos
import net.minecraft.util.math.MathHelper
import net.minecraft.util.math.Vec2f
import net.minecraft.util.math.Vec3d
import java.awt.Color
import java.util.concurrent.CopyOnWriteArrayList
import kotlin.math.floor

/**
 * @author Surge
 * @since 23/08/2022
 */
object Surround : Module("Surround", Category.COMBAT, "Automatically surrounds you with obsidian") {

    private val performOn = Setting("PerformOn", PerformOn.PACKET) describedBy "When to perform"
    private val disable = Setting("Disable", Disable.OFF_GROUND) describedBy "When to automatically disable the module"
    private val center = Setting("Center", Center.MOTION) describedBy "How to center the player to the center of the block"
    private val blocksPerTick = Setting("BlocksPerTick", 4.0, 1.0, 8.0, 1.0) describedBy "The limit to how many blocks can be placed in a tick"
    private val support = Setting("Support", true) describedBy "Support blocks by placing beneath them"

    private val rotate = Setting("Rotate", Rotate.PACKET) describedBy "How to rotate"

    private val render = Setting("Render", true) describedBy "Render a highlight on the positions we need to place blocks at"
    private val renderColour = Setting("Colour", Color(185, 17, 255, 130)) describedBy "The colour of the highlight" subOf render

    // List of positions to place on tick
    private var surroundPositions = arrayListOf<BlockPos>()

    // This is to prevent us from attempting placement multiple times
    private val placedCache = CopyOnWriteArrayList<BlockPos>()

    override fun onEnable() {
        if (minecraft.anyNull) {
            return
        }

        // No obsidian to place
        if (InventoryUtil.getHotbarBlockSlot(Blocks.OBSIDIAN) == -1) {
            Paragon.INSTANCE.notificationManager.addNotification(Notification("No obsidian in hotbar!", NotificationType.ERROR))
            toggle()
            return
        }

        when (center.value) {
            Center.MOTION -> {
                // Set player's motion to walk to the center of the block
                minecraft.player.motionX = (MathHelper.floor(minecraft.player.posX) + 0.5 - minecraft.player.posX) / 2
                minecraft.player.motionZ = (MathHelper.floor(minecraft.player.posZ) + 0.5 - minecraft.player.posZ) / 2
            }

            Center.SNAP -> {
                // Send movement packet
                minecraft.player.connection.sendPacket(CPacketPlayer.Position(MathHelper.floor(minecraft.player.posX) + 0.5, minecraft.player.posY, MathHelper.floor(minecraft.player.posZ) + 0.5, minecraft.player.onGround))

                // Set position client-side
                minecraft.player.setPosition(MathHelper.floor(minecraft.player.posX) + 0.5, minecraft.player.posY, MathHelper.floor(minecraft.player.posZ) + 0.5)
            }

            else -> {}
        }

        surroundPositions = getBlocks()
    }

    override fun onTick() {
        if (minecraft.anyNull) {
            return
        }

        // Remove if non-replaceable block
        placedCache.removeIf { isNotReplaceable(it) }

        // Disable

        // No obsidian to place
        if (InventoryUtil.getHotbarBlockSlot(Blocks.OBSIDIAN) == -1) {
            Paragon.INSTANCE.notificationManager.addNotification(Notification("No obsidian in hotbar!", NotificationType.ERROR))
            toggle()
            return
        }

        if (surroundPositions.isEmpty() && disable.value == Disable.FINISHED) {
            Paragon.INSTANCE.notificationManager.addNotification(Notification("Surround Finished, Disabling!", NotificationType.INFO))
            toggle()
            return
        }

        if (!minecraft.player.onGround && disable.value == Disable.OFF_GROUND) {
            Paragon.INSTANCE.notificationManager.addNotification(Notification("Player is no longer on ground, disabling!", NotificationType.INFO))
            toggle()
            return
        }

        // Refresh blocks on tick
        if (performOn.value == PerformOn.TICK) {
            surroundPositions = getBlocks()
        }

        if (surroundPositions.isNotEmpty()) {
            // List of positions to place at
            val placePositions = ArrayList<BlockPos>()

            // Add all positions from index 0 to the max blocks per tick
            for (i in 0..MathHelper.clamp(surroundPositions.size - 1, 0, blocksPerTick.value.toInt())) {
                placePositions.add(surroundPositions[i])
            }

            // Place blocks
            for (pos in placePositions) {
                placeOnPosition(pos)
            }

            // Remove placed positions from global list
            surroundPositions.removeAll(placePositions.toSet())
        }
    }

    @Listener
    fun onPacketReceived(event: PacketEvent.PreReceive) {
        if (event.packet is SPacketBlockChange && performOn.value == PerformOn.PACKET) {
            val pos = event.packet.blockPosition

            // It's a placeable position, and we haven't already attempted to place there
            if (!isNotReplaceable(pos) && !placedCache.contains(pos)) {
                // Check sub (support) status
                val sub = if (support.value && pos.down().getBlockAtPos().isReplaceable(minecraft.world, pos.down())) pos.down() else null

                // Place sub block
                if (sub != null) {
                    placeOnPosition(sub)
                }

                // Place origin block
                placeOnPosition(pos)

                // Otherwise we seem to attempt placing 2-3 times
                placedCache.add(pos)
            }
        }
    }

    override fun onRender3D() {
        surroundPositions.forEach {
            RenderBuilder()
                .boundingBox(BlockUtil.getBlockBox(it))
                .inner(renderColour.value)
                .outer(renderColour.value.integrateAlpha(255f))

                .start()

                .blend(true)
                .texture(true)
                .depth(true)

                .build(false)
        }
    }

    private fun isNotReplaceable(pos: BlockPos): Boolean = !pos.getBlockAtPos().blockState.block.isReplaceable(minecraft.world, pos)

    private fun getBlocks(origin: BlockPos): ArrayList<BlockPos> {
        val blocks = arrayListOf<BlockPos>()

        // If we don't want to support, ignore this block
        if (!support.value && origin.getBlockAtPos().isReplaceable(minecraft.world, origin) && origin.down().getBlockAtPos().isReplaceable(minecraft.world, origin.down())) {
            return blocks
        }

        // Check that the origin block is replaceable
        if (origin.getBlockAtPos().isReplaceable(minecraft.world, origin)) {
            // If we want to support, and the block below is replaceable, add it to the list
            if (support.value && origin.down().getBlockAtPos().isReplaceable(minecraft.world, origin.down())) {
                blocks.add(origin.down())
            }

            // Add origin
            blocks.add(origin)
        }

        return blocks
    }

    private fun getBlocks(): ArrayList<BlockPos> {
        val blocks = ArrayList<BlockPos>()
        val playerPos = BlockPos(floor(minecraft.player.posX), floor(minecraft.player.posY), floor(minecraft.player.posZ)).add(0, 1, 0)

        // Add blocks
        blocks.addAll(getBlocks(playerPos.add(-1, -1, 0)))
        blocks.addAll(getBlocks(playerPos.add(1, -1, 0)))
        blocks.addAll(getBlocks(playerPos.add(0, -1, -1)))
        blocks.addAll(getBlocks(playerPos.add(0, -1, 1)))

        return blocks
    }

    private fun placeOnPosition(position: BlockPos) {
        // Get current item
        val slot: Int = minecraft.player.inventory.currentItem

        // Slot to switch to
        val obsidianSlot = InventoryUtil.getHotbarBlockSlot(Blocks.OBSIDIAN)

        if (obsidianSlot != -1) {
            minecraft.player.inventory.currentItem = obsidianSlot

            // Get rotation yaw and pitch
            val rotationValues = RotationUtil.getRotationToBlockPos(position, 0.5)

            // Place
            PlacementUtil.place(position, Rotation(rotationValues.x, rotationValues.y, rotate.value, RotationPriority.HIGH))

            // Reset slot to our original slot
            minecraft.player.inventory.currentItem = slot
        }
    }

    enum class PerformOn {
        /**
         * Place when block is destroyed
         */
        PACKET,

        /**
         * Place on tick
         */
        TICK
    }

    enum class Disable {
        /**
         * Disable when finished
         */
        FINISHED,

        /**
         * Disable when off ground
         */
        OFF_GROUND,

        /**
         * Never disable
         */
        NEVER
    }

    enum class Center {
        /**
         * Move the player to the center of the block
         */
        MOTION,

        /**
         * Snap the player to the center of the block
         */
        SNAP,

        /**
         * Do not center the player
         */
        OFF
    }

}